### PR TITLE
[WIP] Migration for settings

### PR DESCRIPTION
**Pull request by allouis**
The 2.17 migration included a bug which set the `is_private`, `amp` and `force_i18n` setting values to `'false'` when they should have been `'true'`

We've reverted these changes by reading the most recent backup file, and setting the value to `'true'` if the backup has it set to `'true'` AND the current db has it set to false.

We've also updated the naming for database backups to include the timestamp, this is so that multiple backups in one day will not overwrite each other.

We've also deleted the broken migration, so it will not run on future installs.

### CODE REVIEW

File: core/server/data/exporter/index.js [hunk 1]
```diff
@@ -25,7 +25,7 @@ var _ = require('lodash'),
     exportFileName;
 
 exportFileName = function exportFileName(options) {
-    var datetime = (new Date()).toJSON().substring(0, 10),
+    var datetime = require('moment')().format('YYYY-MM-DD-HH-mm-ss'),
```
Comment (allouis) [inline core/server/data/exporter/index.js]: This is so backups on the same day don't overwrite

File: core/server/data/migrations/versions/2.18/1-restore-settings-from-backup.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+const common = require('../../../../lib/common');
+const config = require('../../../../config');
+const fs = require('fs-extra');
+const path = require('path');
+
+module.exports.config = {
+    transaction: true
+};
+
+const backupFileRegex = /ghost.([\d]{4}-[\d]{2}-[\d]{2}).json$/;
```
Comment (allouis) [inline core/server/data/migrations/versions/2.18/1-restore-settings-from-backup.js]: Only matches "old" backup filename (Pre- these changes)

File: core/server/data/migrations/versions/2.18/1-restore-settings-from-backup.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        const backups = files.filter(function (filename) {
+            return backupFileRegex.test(filename);
+        }).sort(function (a, b) {
+            const dateA = new Date(a.match(backupFileRegex)[1]);
+            const dateB = new Date(b.match(backupFileRegex)[1]);
+
+            return dateB - dateA;
+        });
+
+        const mostRecentBackup = backups[0];
```
Comment (allouis) [inline core/server/data/migrations/versions/2.18/1-restore-settings-from-backup.js]: This is most recent backup of the old style (pre these changes)

File: core/server/data/migrations/versions/2.18/1-restore-settings-from-backup.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            throw new Error('Could not read settings from backup file');
+        }
+
+        const relevantBackupSettings = settings.filter(function (entry) {
+            return ['is_private', 'force_i18n', 'amp'].includes(entry.key);
+        }).reduce(function (obj, entry) {
+            return Object.assign(obj, {
+                [entry.key]: entry
+            });
+        }, {});
```
Comment (allouis) [inline core/server/data/migrations/versions/2.18/1-restore-settings-from-backup.js]: Results in object like:
```
{
  "is_private": <settingObjectFor is_private>,
  "amp": <settingObjectFor amp>,
 etc..
}
```

**Comment by allouis** [comment]
I don't think we need to consider a `down` script for this?

**Comment by allouis** [comment]
This PR should delete the broken migration too

File: core/server/data/migrations/versions/2.18/1-restore-settings-from-backup.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                    return;
+                }
+
+                // @NOTE: ensure we only transform real boolean fields to ensure we won't modify any customer data
+                const relevantLiveSettings = response.filter(function (entry) {
+                    return ['is_private', 'force_i18n', 'amp'].includes(entry.key);
+                });
+
+                return Promise.each(relevantLiveSettings, (liveSetting) => {
+                    // @NOTE: Something else is stored (any other value, set to false), normalize boolean fields
```
Comment (kevinansfield) [inline core/server/data/migrations/versions/2.18/1-restore-settings-from-backup.js]: @allouis what is this comment referring to?

Comment (allouis) [reply core/server/data/migrations/versions/2.18/1-restore-settings-from-backup.js]: ^ that is old - i'm sorry, will delete

**Pull request merged**